File: backend/internal/domain/card_subtype.go
package domain

import "strconv"

// CardSubtype identifies a card subtype token (persisted as smallint in cards.subtypes).
// Token strings match the legacy TypeScript enum values ("1H", "Non-Attack", …).
// Values must stay in sync with frontend `src/constants/cardSubtype.js`.
type CardSubtype int16

const (
	CardSubtypeOneHanded CardSubtype = iota
	CardSubtypeTwoHanded
	CardSubtypeAffliction
	CardSubtypeAlly
	CardSubtypeAngel
	CardSubtypeArms
	CardSubtypeArrow
	CardSubtypeAsh
	CardSubtypeAttack
	CardSubtypeAura
	CardSubtypeAxe
	CardSubtypeBase
	CardSubtypeBook
	CardSubtypeBow
	CardSubtypeCannon
	CardSubtypeChest
	CardSubtypeChi
	CardSubtypeClaw
	CardSubtypeClub
	CardSubtypeConstruct
	CardSubtypeDagger
	CardSubtypeDemon
	CardSubtypeDragon
	CardSubtypeEvo
	CardSubtypeFiddle
	CardSubtypeFigment
	CardSubtypeFlail
	CardSubtypeGem
	CardSubtypeGun
	CardSubtypeHammer
	CardSubtypeHead
	CardSubtypeInvocation
	CardSubtypeItem
	CardSubtypeLandmark
	CardSubtypeLog
	CardSubtypeLute
	CardSubtypeLegs
	CardSubtypeNonAttack
	CardSubtypeOffHand
	CardSubtypeOrb
	CardSubtypePistol
	CardSubtypePitFighter
	CardSubtypePolearm
	CardSubtypeQuiver
	CardSubtypeRock
	CardSubtypeShuriken
	CardSubtypeScepter
	CardSubtypeScroll
	CardSubtypeScythe
	CardSubtypeSong
	CardSubtypeStaff
	CardSubtypeSword
	CardSubtypeTrap
	CardSubtypeWrench
	CardSubtypeYoung
)

var cardSubtypeNames = map[CardSubtype]string{
	CardSubtypeOneHanded:  "1H",
	CardSubtypeTwoHanded:  "2H",
	CardSubtypeAffliction: "Affliction",
	CardSubtypeAlly:       "Ally",
	CardSubtypeAngel:      "Angel",
	CardSubtypeArms:       "Arms",
	CardSubtypeArrow:      "Arrow",
	CardSubtypeAsh:        "Ash",
	CardSubtypeAttack:     "Attack",
	CardSubtypeAura:       "Aura",
	CardSubtypeAxe:        "Axe",
	CardSubtypeBase:       "Base",
	CardSubtypeBook:       "Book",
	CardSubtypeBow:        "Bow",
	CardSubtypeCannon:     "Cannon",
	CardSubtypeChest:      "Chest",
	CardSubtypeChi:        "Chi",
	CardSubtypeClaw:       "Claw",
	CardSubtypeClub:       "Club",
	CardSubtypeConstruct:  "Construct",
	CardSubtypeDagger:     "Dagger",
	CardSubtypeDemon:      "Demon",
	CardSubtypeDragon:     "Dragon",
	CardSubtypeEvo:        "Evo",
	CardSubtypeFiddle:     "Fiddle",
	CardSubtypeFigment:    "Figment",
	CardSubtypeFlail:      "Flail",
	CardSubtypeGem:        "Gem",
	CardSubtypeGun:        "Gun",
	CardSubtypeHammer:     "Hammer",
	CardSubtypeHead:       "Head",
	CardSubtypeInvocation: "Invocation",
	CardSubtypeItem:       "Item",
	CardSubtypeLandmark:   "Landmark",
	CardSubtypeLog:        "Log",
	CardSubtypeLute:       "Lute",
	CardSubtypeLegs:       "Legs",
	CardSubtypeNonAttack:  "Non-Attack",
	CardSubtypeOffHand:    "Off-Hand",
	CardSubtypeOrb:        "Orb",
	CardSubtypePistol:     "Pistol",
	CardSubtypePitFighter: "Pit-Fighter",
	CardSubtypePolearm:    "Polearm",
	CardSubtypeQuiver:     "Quiver",
	CardSubtypeRock:       "Rock",
	CardSubtypeShuriken:   "Shuriken",
	CardSubtypeScepter:    "Scepter",
	CardSubtypeScroll:     "Scroll",
	CardSubtypeScythe:     "Scythe",
	CardSubtypeSong:       "Song",
	CardSubtypeStaff:      "Staff",
	CardSubtypeSword:      "Sword",
	CardSubtypeTrap:       "Trap",
	CardSubtypeWrench:     "Wrench",
	CardSubtypeYoung:      "Young",
}

var cardSubtypeByToken map[string]CardSubtype

func init() {
	cardSubtypeByToken = make(map[string]CardSubtype, len(cardSubtypeNames))
	for id, tok := range cardSubtypeNames {
		cardSubtypeByToken[tok] = id
	}
}

// Valid reports whether s is a defined CardSubtype constant.
func (s CardSubtype) Valid() bool {
	_, ok := cardSubtypeNames[s]
	return ok
}

// Token returns the legacy subtype string ("1H", "Non-Attack", …) or "" if unknown.
func (s CardSubtype) Token() string {
	return cardSubtypeNames[s]
}

// String satisfies fmt.Stringer; echoes Token() for defined values.
func (s CardSubtype) String() string {
	if t := s.Token(); t != "" {
		return t
	}
	return "CardSubtype(" + strconv.FormatInt(int64(s), 10) + ")"
}

// CardSubtypeFromToken maps an import/display token to its numeric ID (e.g. "1H", "Non-Attack").
func CardSubtypeFromToken(tok string) (CardSubtype, bool) {
	sub, ok := cardSubtypeByToken[tok]
	return sub, ok
}

// CardSubtypes returns IDs 0..CardSubtypeYoung in order.
func CardSubtypes() []CardSubtype {
	out := make([]CardSubtype, 0, len(cardSubtypeNames))
	for s := CardSubtype(0); s <= CardSubtypeYoung; s++ {
		out = append(out, s)
	}
	return out
}

File: src/constants/cardSubtype.js
/**
 * Card subtype enum IDs (smallint in `subtypes` arrays).
 * Matches `backend/internal/domain/card_subtype.go` and legacy TS `Subtype` string values (`CardSubtype.OneHanded` → `"1H"` via `cardSubtypeToken(0)`).
 * @readonly
 */
export const CardSubtype = Object.freeze({
  OneHanded: 0,
  TwoHanded: 1,
  Affliction: 2,
  Ally: 3,
  Angel: 4,
  Arms: 5,
  Arrow: 6,
  Ash: 7,
  Attack: 8,
  Aura: 9,
  Axe: 10,
  Base: 11,
  Book: 12,
  Bow: 13,
  Cannon: 14,
  Chest: 15,
  Chi: 16,
  Claw: 17,
  Club: 18,
  Construct: 19,
  Dagger: 20,
  Demon: 21,
  Dragon: 22,
  Evo: 23,
  Fiddle: 24,
  Figment: 25,
  Flail: 26,
  Gem: 27,
  Gun: 28,
  Hammer: 29,
  Head: 30,
  Invocation: 31,
  Item: 32,
  Landmark: 33,
  Log: 34,
  Lute: 35,
  Legs: 36,
  NonAttack: 37,
  OffHand: 38,
  Orb: 39,
  Pistol: 40,
  PitFighter: 41,
  Polearm: 42,
  Quiver: 43,
  Rock: 44,
  Shuriken: 45,
  Scepter: 46,
  Scroll: 47,
  Scythe: 48,
  Song: 49,
  Staff: 50,
  Sword: 51,
  Trap: 52,
  Wrench: 53,
  Young: 54,
});

/** Legacy string token for each ID 0..54 (`Subtype` enum values in TypeScript). */
export const CARD_SUBTYPE_TOKENS = Object.freeze([
  "1H",
  "2H",
  "Affliction",
  "Ally",
  "Angel",
  "Arms",
  "Arrow",
  "Ash",
  "Attack",
  "Aura",
  "Axe",
  "Base",
  "Book",
  "Bow",
  "Cannon",
  "Chest",
  "Chi",
  "Claw",
  "Club",
  "Construct",
  "Dagger",
  "Demon",
  "Dragon",
  "Evo",
  "Fiddle",
  "Figment",
  "Flail",
  "Gem",
  "Gun",
  "Hammer",
  "Head",
  "Invocation",
  "Item",
  "Landmark",
  "Log",
  "Lute",
  "Legs",
  "Non-Attack",
  "Off-Hand",
  "Orb",
  "Pistol",
  "Pit-Fighter",
  "Polearm",
  "Quiver",
  "Rock",
  "Shuriken",
  "Scepter",
  "Scroll",
  "Scythe",
  "Song",
  "Staff",
  "Sword",
  "Trap",
  "Wrench",
  "Young",
]);

/** @readonly */
export const CARD_SUBTYPE_ID_BY_TOKEN = Object.freeze(
  Object.fromEntries(CARD_SUBTYPE_TOKENS.map((t, i) => [t, i])),
);

/**
 * @param {number} id
 * @returns {string | undefined}
 */
export function cardSubtypeToken(id) {
  return CARD_SUBTYPE_TOKENS[id];
}

/**
 * @param {string} token legacy subtype string, e.g. `"1H"` or `"Non-Attack"`
 * @returns {number | undefined}
 */
export function cardSubtypeFromToken(token) {
  const n = CARD_SUBTYPE_ID_BY_TOKEN[token];
  return typeof n === "number" ? n : undefined;
}

/**
 * @param {number} id
 * @returns {boolean}
 */
export function isValidCardSubtypeId(id) {
  return Number.isInteger(id) && id >= 0 && id < CARD_SUBTYPE_TOKENS.length;
}
